### PR TITLE
Death Logs

### PR DESCRIPTION
**Pull request by Nick42791**
---Defeat Screen---
*Has logs of what happened during the game

*New Defeat logic, which should fix Sandstorm error

---Tutorial Changes---
*Upkeep renamed to system health in investment card

*Trade Request amount variables are set to 0 when reset

*Trades no longer have a blue bar between them

*Trade buttons expand in the x direction, but not y.

*Tutorial Quiz incorrect/correct style handling

### CODE REVIEW

**Comment by alee** [comment]
this is failing the travis build due to a typescript issue:

```
ERROR in /code/client/src/components/gamedashboard/global/cards/CardInvestment.vue(84,12):
35584:12 This condition will always return 'false' since the types '"science" | "government" | "legacy" | "finance" | "culture"' and '"upkeep"' have no overlap.
356    82 | 
357    83 |   get returnName(){
358  > 84 |     return this.name == 'upkeep' ? 'System Health' : this.name;
359       |            ^
360    85 |   }
361    86 | 
362    87 |   private setInvestmentAmount(diff: number): void {
363
364 ERROR  Build failed with errors.
```

File: client/src/components/gamedashboard/top/MarsLog.vue [hunk 1]
```diff
@@ -36,7 +36,9 @@ export default class MarsLog extends Vue {
   marsLogColor(log: MarsLogData) {
     console.log(log.category)
     switch (log.category) {
-      case 'upkeep':
+      case 'System Health: Gain':
```
Comment (alee) [inline client/src/components/gamedashboard/top/MarsLog.vue]: instead of explicit `System Health: Gain` and `System Health: Drop` categories can we base this logic on whether the payload integer value is positive or negative?

File: client/src/components/gamedashboard/global/containers/ContainerDefeat.vue [hunk 2]
```diff
@@ -1,23 +1,34 @@
...
-    <div class="text">
-      <p>
-        System Health has reached zero. However, your efforts will inspire a new
-        generation to continue your legacy on Mars.
-      </p>
-      <p>Thank you for playing.</p>
+    <h1 class="animated pulse slower infinite gameOver-text">Game Over</h1>
+    <div class="transmission-info">
+      <h3>Replaying Logs before death...</h3>
+
```
Comment (alee) [inline client/src/components/gamedashboard/global/containers/ContainerDefeat.vue]: We probably want to keep the `System Health has reached zero...` text (or something like it) here to keep the narrative in addition to the pile of log events.

File: server/src/rooms/game/commands/index.ts [hunk 3]
```diff
@@ -215,41 +219,44 @@ export class SetNextPhaseCmd implements Command {
...
       case Phase.invest:
         return [new EnteredTradePhase()];
       case Phase.trade:
         return [new EnteredPurchasePhase()];
       case Phase.purchase:
-        this.gameOver(this.upkeep);
-        return [new EnteredDiscardPhase()];
+        
+        
+        return [new EnteredDiscardPhase(), new EnteredDefeatPhase()];
```
Comment (alee) [inline server/src/rooms/game/commands/index.ts]: can you explain what's going on here? why are we adding the `EnteredDefeatPhase` here?

**Comment by chrstngyn** [comment]
Logs should read "system health" not upkeep

File: server/src/rooms/game/events/index.ts [hunk 4]
```diff
@@ -345,7 +357,18 @@ gameEventDeserializer.register(EnteredDiscardPhase);
 
 export class EnteredDefeatPhase extends KindOnlyGameEvent {
   apply(game: GameState): void {
-    game.phase = Phase.defeat;
+    if(game.upkeep <= 0){
```
Comment (alee) [inline server/src/rooms/game/events/index.ts]: I'm not sure I'm comfortable with always adding the defeat phase and checking inside it to see if we should be in the defeat phase.. perhaps this is a naming issue and I'd be fine if it were called `CleanupPhase` or `VerifyPhase` or something along those lines. Other thoughts?

File: server/src/rooms/game/events/index.ts [hunk 5]
```diff
@@ -345,7 +357,18 @@ gameEventDeserializer.register(EnteredDiscardPhase);
...
+      game.phase = Phase.defeat;
+      const log = new MarsLogMessage({
+        performedBy: 'Server',
+        category: 'System Health',
+        content: `System health has reached zero.`,
+        timestamp: this.dateCreated
+      });
+
+      game.logs.push(log);
```
Comment (alee) [inline server/src/rooms/game/events/index.ts]: use `game.log(..)` instead of manual log construction

File: client/src/components/gamedashboard/global/cards/CardInvestment.vue [hunk 6]
```diff
@@ -2,7 +2,7 @@
   <div class="card-investment v-step-15">
     <div class="ci-container" :style="opacity">
       <div class="type">
-        <p class="name">{{ name }}</p>
+        <p class="name">{{ returnName }}</p>
```
Comment (alee) [inline client/src/components/gamedashboard/global/cards/CardInvestment.vue]: minor nitpick, I think this should probably be renamed to `label`. `returnName` isn't a super helpful variable name :sweat_smile:

File: client/src/components/gamedashboard/top/MarsLog.vue (hunk 1 above)
Comment (alee) [reply client/src/components/gamedashboard/top/MarsLog.vue]: disregard this comment, I realized you're doing this off the logs which don't have that content.

File: server/src/rooms/game/events/index.ts (hunk 4 above)
Comment (Nick42791) [reply server/src/rooms/game/events/index.ts]: So the reason for this is because of the way the events get applied. The finalize methods and the defeat methods are applied in the same sweep. So, for example, sandstorm drops upkeep in its finalize method, but the defeat command is evaluated before the drop is applied, making it possible for upkeep to be negative. By always applying defeat and moving the logic to the event, we can always get the most recent upkeep, so it works. I hope that’s clear, but if it’s not @cpritcha might be able to explain better.

File: client/src/components/gamedashboard/global/containers/ContainerDefeat.vue (hunk 2 above)
Comment (Nick42791) [reply client/src/components/gamedashboard/global/containers/ContainerDefeat.vue]: The last log that comes in says that system health has reached zero. We can re-iterate it if you want, but it is there.

File: server/src/rooms/game/events/index.ts (hunk 4 above)
Comment (alee) [reply server/src/rooms/game/events/index.ts]: My concern has more to do with naming and clarity of the codebase. We're not entering the defeat phase, we are checking whether the group is still alive and kicking.

I'd like for us to have ELI5-able semantics on how we are applying system health contributions and decrements and when they are checked. @cpritcha @chrstngyn feel free to chime in though if we want to punt on this for tomorrow's pretest since the other changes look good to me and are a clear improvement over the nothing that's there now.

File: client/src/components/gamedashboard/global/containers/ContainerDefeat.vue (hunk 2 above)
Comment (alee) [reply client/src/components/gamedashboard/global/containers/ContainerDefeat.vue]: There needs to be a narrative goodbye, which I don't expect to exist in the mars log. Don't worry too much about this, I can add it in later.

**Comment by alee** [comment]
hey Nick, next time please do PRs from a feature branch on your fork, otherwise we'll be stomping on your fork's master if we want to commit to it.
